### PR TITLE
Create hook for post_load data transforms

### PR DESCRIPTION
**Pull request by diehlbw**
# Overview
Added a hook to DataLoader (and loader_factory) to accept a ConfigFrameHook function; allowing modification of the Seismogram dataframe 'during' load.  
This necessitated improving the separation of config and Seismogram, slimming the constructor of the latter and moving a couple configuration manipulations to ConfigProvider

## Review focus
- [ ] verify existing example notebooks work without modification
- [ ] use the new documentation (under userguide) to create a custom function and instantiate a Seismogram using it

## Author Checklist
- [ ] Linting passes; run early with [pre-commit hook](https://pre-commit.com/#install).
- [ ] Tests added for new code and issue being fixed.
- [ ] Added type annotations and full numpy-style docstrings for new methods.
- [ ] Draft your news fragment in new `changelog/ISSUE.TYPE.rst` files; see [changelog/README.md](https://github.com/epic-open-source/seismometer/blob/main/changelog/README.md).

### CODE REVIEW

File: docs/user_guide/index.rst [hunk 1]
```diff
@@ -502,3 +502,29 @@ the model's name and any configured thresholds. It is typically defined in a
       "modelname": "Risk of Readmission for Patients with Diabetes",
       "thresholds": [0.65, 0.3]
    }
+
+Modifying the Analysis Data
+---------------------------
+
+When possible it can be better to modify the data during extraction and have predictions and events files contain everything that is needed for analysis.
+However, there will inevitably be times when this is not possible but you still want transformations to be done prior to most of the notebook running.
```
Comment (gbowlin) [inline docs/user_guide/index.rst:510]:
```suggestion
When possible it is better to modify the data during extraction and have both predictions and events files contain everything that is needed for analysis. 
Inevitably, there will be times when this is not possible and you need additional transformations to be done prior to most of the notebook running.
```

File: docs/user_guide/index.rst [hunk 2]
```diff
@@ -502,3 +502,29 @@ the model's name and any configured thresholds. It is typically defined in a
...
+First, create a :py:func:`~seismometer.data.loader.ConfigFrameHook` (accepts a :py:class:`~seismometer.configuration.config.ConfigProvider`) and a pd.DataFrame and outputs a pd.DataFrame) that can modify the standard Seismogram frame to the desired state.
+Then, follow the pattern of normal startup but specify your function in the :py:func:`~seismometer.data.loader.loader_factory`:
+
+.. code-block:: python
+
+   from seismometer.configuration import ConfigProvider
+   from seismometer.data.loader import loader_factory
+   from seismometer.seismogram import Seismogram
+
+   def my_startup(config_path="."):
```
Comment (gbowlin) [inline docs/user_guide/index.rst:523]: Do they need to setup a logger?

File: src/seismometer/__init__.py [hunk 3]
```diff
@@ -44,6 +44,8 @@ def run_startup(
     """
     import importlib
 
+    from seismometer.configuration import ConfigProvider
+    from seismometer.data.loader import loader_factory
     from seismometer.seismogram import Seismogram
 
     set_default_logger_config()
```
Comment (gbowlin) [inline src/seismometer/__init__.py:51]: How will custom startups handling logger creation?

File: src/seismometer/seismogram.py [hunk 4]
```diff
@@ -61,29 +57,19 @@ def __init__(
...
         self.dataframe: pd.DataFrame = None
         self.cohort_cols: list[str] = []
-        self.config_path = config_path
-
-        self.load_config(config_path, definitions=definitions)
 
-        self.config.set_output(output_path)
-        self.config.output_dir.mkdir(parents=True, exist_ok=True)
-        self.dataloader = loader_factory(self.config)
+        self.copy_config_metadata()
```
Comment (gbowlin) [inline src/seismometer/seismogram.py:72]: should we copy the config metadata, or just have properties fall through to metadata?

**Review by gbowlin — COMMENTED** [review]
Need to do functional testing, but otherwise looks fine. I like the constructor cleanup to only need Config +  DataLoader.

File: docs/user_guide/index.rst (hunk 2 above)
Comment (gbowlin) [reply docs/user_guide/index.rst:523]: Also do they need to do anything to make sure `_api` is in the exposed as part of the `seismometer` namespace

File: src/seismometer/data/loader/pipeline.py [hunk 5]
```diff
@@ -65,6 +66,7 @@ def __init__(
         post_predict_fn: Optional[ConfigFrameHook] = None,
         post_event_fn: Optional[ConfigFrameHook] = None,
         merge_fn: Optional[MergeFramesHook] = None,
+        post_load_fn: Optional[ConfigFrameHook] = None,
```
Comment (firste) [inline src/seismometer/data/loader/pipeline.py:69]: Does it also make sense to do some validation on the code hook itself? Pydantic appears to have some support for this already -  https://docs.pydantic.dev/latest/api/validate_call/#pydantic.validate_call_decorator.validate_call

**Review by firste — CHANGES_REQUESTED** [review]

File: docs/user_guide/index.rst [hunk 6]
```diff
@@ -502,3 +502,29 @@ the model's name and any configured thresholds. It is typically defined in a
...
+The general outline of what the code should do is the same but will take advantage of the post_load_fn hook.
+First, create a :py:func:`~seismometer.data.loader.ConfigFrameHook` (accepts a :py:class:`~seismometer.configuration.config.ConfigProvider`) and a pd.DataFrame and outputs a pd.DataFrame) that can modify the standard Seismogram frame to the desired state.
+Then, follow the pattern of normal startup but specify your function in the :py:func:`~seismometer.data.loader.loader_factory`:
+
+.. code-block:: python
+
+   from seismometer.configuration import ConfigProvider
+   from seismometer.data.loader import loader_factory
+   from seismometer.seismogram import Seismogram
+
```
Comment (gbowlin) [inline docs/user_guide/index.rst:522]:
```suggestion
   from seismometer.seismogram import Seismogram
   import seismometer._api as sm
```

File: src/seismometer/data/loader/pipeline.py (hunk 5 above)
Comment (diehlbw) [reply src/seismometer/data/loader/pipeline.py:69]: Played with this a bit, and think I want to push this kind of validation assistance to the larger planned RFC around more config-based support for post_load transforms.

The validate_call is kind of the opposite direction than we want, and to help we'd have to 1-require that post_load_fn be defined with annotations and then do inspection on the signature.

**Review by gbowlin — APPROVED** [review]

**Review by firste — APPROVED** [review]

**Pull request merged**